src/Castle.Core.Tests/DynamicProxy.Tests/InParamsTestCase.cs

```diff
@@ -0,0 +1,165 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Castle.DynamicProxy.Tests
+{
+	using System;
+	using System.Linq;
+	using System.Reflection;
+
+	using NUnit.Framework;
+
+	[TestFixture]
+	public class InParamsTestCase : BasePEVerifyTestCase
+	{
+		[Test]
+		public void By_value_parameter_cannot_modify_argument_var()
+		{
+			var x = new ReadOnlyStruct(1);
+			var original = new ReadOnlyStruct(x.Value);
+
+			var different = new ReadOnlyStruct(x.Value + 100);
+			var proxy = generator.CreateInterfaceProxyWithoutTarget<IByValue>(new SetArgumentValueInterceptor(different));
+			proxy.Method(x);
+
+			Assert.AreEqual(original.Value, x.Value);
+		}
+
+		[Test]
+		public void By_referece_In_parameter_cannot_modify_argument_var()
+		{
+			var x = new ReadOnlyStruct(1);
+			var original = new ReadOnlyStruct(x.Value);
+
+			var different = new ReadOnlyStruct(x.Value + 100);
+			var proxy = generator.CreateInterfaceProxyWithoutTarget<IByReadOnlyRef>(new SetArgumentValueInterceptor(different));
+			proxy.Method(in x);
+
+			Assert.AreEqual(original.Value, x.Value);
+		}
+
+		[Test]
+		public void By_reference_Ref_parameter_can_modify_argument_var()
+		{
+			var x = new ReadOnlyStruct(1);
+			var original = new ReadOnlyStruct(x.Value);
+
+			var different = new ReadOnlyStruct(x.Value + 100);
+			var proxy = generator.CreateInterfaceProxyWithoutTarget<IByRef>(new SetArgumentValueInterceptor(different));
+			proxy.Method(ref x);
+
+			Assert.AreNotEqual(original.Value, x.Value);
+		}
+
+		[Test]
+		public void By_reference_Out_parameter_can_modify_argument_var()
+		{
+			var x = new ReadOnlyStruct(1);
+			var original = new ReadOnlyStruct(x.Value);
+
+			var different = new ReadOnlyStruct(x.Value + 100);
+			var proxy = generator.CreateInterfaceProxyWithoutTarget<IOut>(new SetArgumentValueInterceptor(different));
+			proxy.Method(out x);
+
+			Assert.AreNotEqual(original.Value, x.Value);
+		}
+
+		[Test]
+		[TestCase(typeof(IByReadOnlyRef))]
+		public void By_reference_In_parameter_has_ParameterAttributes_In_set(Type type)
+		{
+			var parameter = type.GetMethod("Method").GetParameters()[0];
+
+			Assert.True(parameter.Attributes.HasFlag(ParameterAttributes.In));
+		}
+
+		[Test]
+		[TestCase(typeof(IByValue))]
+		[TestCase(typeof(IByRef))]
+		[TestCase(typeof(IOut))]
+		public void Parameter_other_than_by_reference_In_does_not_have_ParameterAttributes_In_set(Type type)
+		{
+			var parameter = type.GetMethod("Method").GetParameters()[0];
+
+			Assert.False(parameter.Attributes.HasFlag(ParameterAttributes.In));
+		}
+
+		[Test]
+		[TestCase(typeof(IByReadOnlyRef))]
+		public void By_reference_In_parameter_has_IsReadOnlyAttribute(Type type)
+		{
+			var parameter = type.GetMethod("Method").GetParameters()[0];
+
+			Assert.True(parameter.GetCustomAttributes().Any(a => a.GetType().FullName == "System.Runtime.CompilerServices.IsReadOnlyAttribute"));
```

Comment:
I thought these classes were public classes, I assume they aren't visible in .NET Standard 1.3?